### PR TITLE
iOS image picker

### PR DESCRIPTION
**Pull request by sklencar**
Enable pick image from a gallery option for iOS

### CODE REVIEW

File: app/inpututils.h [hunk 1]
```diff
@@ -45,6 +45,8 @@ class InputUtils: public QObject
      */
     Q_INVOKABLE QVector<double> extractGeometryCoordinates( const QgsQuickFeatureLayerPair &pair, QgsQuickMapSettings *mapSettings );
 
+    Q_INVOKABLE QString extractRelativePath( const QString &path, const QString &prefixPath );
```
Comment (PeterPetrik) [inline app/inpututils.h]: add documentation

File: app/ios/iosdevice.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
+#include <QPointer>
```
Comment (PeterPetrik) [inline app/ios/iosdevice.cpp]: - add copyright header
- add to ios.pri sources?

File: app/ios/iosdevice.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+QVariantMap QIDevice::fetch() const
+{
+  QVariantMap data;
+  return data;
+}
+#endif
+
+
+QIDevice::~QIDevice()
+{
```
Comment (PeterPetrik) [inline app/ios/iosdevice.cpp]: = default?

File: app/ios/iosdevice.cpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  if ( m_instance.isNull() )
+  {
+    QCoreApplication *app = QCoreApplication::instance();
+    m_instance = new QIDevice( app );
+  }
+
+  return m_instance;
+}
+
+
```
Comment (PeterPetrik) [inline app/ios/iosdevice.cpp]: too many blank lines?

File: app/ios/iosdevice.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
+#ifndef QIDEVICE_H
```
Comment (PeterPetrik) [inline app/ios/iosdevice.h]: - add copyright header
- add to ios.pri sources?

File: app/ios/iosdevice.h [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+#define QIDEVICE_H
+
+#include <QObject>
+#include <QVariantMap>
+
+/// Device information provider
+
+class QIDevice : public QObject
```
Comment (PeterPetrik) [inline app/ios/iosdevice.h]: the name suggests it is part of QT framework, maybe some more "Input"-app prefix?

File: app/ios/iosdevice.h [hunk 7]
```diff
@@ -0,0 +1,55 @@
+#ifndef QIDEVICE_H
+#define QIDEVICE_H
+
+#include <QObject>
+#include <QVariantMap>
+
+/// Device information provider
```
Comment (PeterPetrik) [inline app/ios/iosdevice.h]: use doxy comments. add documentation what this class is responsible for

File: app/ios/iosdevice.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    Q_OBJECT
+    Q_PROPERTY( bool screenFillStatusBar READ screenFillStatusBar WRITE setScreenFillStatusBar NOTIFY screenFillStatusBarChanged )
+    Q_PROPERTY( int screenWidth READ screenWidth WRITE setScreenWidth NOTIFY screenWidthChanged )
+    Q_PROPERTY( int screenHeight READ screenHeight WRITE setScreenHeight NOTIFY screenHeightChanged )
+    Q_PROPERTY( QString identifierForVendor READ identifierForVendor WRITE setIdentifierForVendor NOTIFY identifierForVendorChanged )
+
+  public:
+    ~QIDevice();
+
+    /// It is truth if the status bar area should be filled by the application
```
Comment (PeterPetrik) [inline app/ios/iosdevice.h]: use doxy comment style

File: app/ios/iosdevice.mm [hunk 9]
```diff
@@ -0,0 +1,14 @@
+#import <UIKit/UIKit.h>
```
Comment (PeterPetrik) [inline app/ios/iosdevice.mm]: - add copyright header

File: app/ios/iosdevice.mm [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+    UIDevice* device = [UIDevice currentDevice];
+
+    QVariantMap result;
+
+    result["identifierForVendor"] = QString::fromNSString([[device identifierForVendor] UUIDString]);
+
+    return result;
+}
+
+
```
Comment (PeterPetrik) [inline app/ios/iosdevice.mm]: extra newline?

File: app/ios/iosimagepicker.cpp [hunk 11]
```diff
@@ -0,0 +1,265 @@
+#include <QStandardPaths>
```
Comment (PeterPetrik) [inline app/ios/iosimagepicker.cpp]: - add copyright header
- add to ios.pri sources?

File: app/ios/iosimagepicker.cpp [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+#include <QImageWriter>
+#include <QImageReader>
+#include "iossystemdispatcher.h"
+#include "iosimagepicker.h"
+#include <QtDebug>
+#include <QCoreApplication>
+
+#include "inpututils.h"
+
+class IOSImagePickerSaver : public QRunnable
```
Comment (PeterPetrik) [inline app/ios/iosimagepicker.cpp]: add docs what this class is responsible for

File: app/ios/iosimagepicker.cpp [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+{
+  public:
+    QPointer<IOSImagePicker> owner;
```
Comment (PeterPetrik) [inline app/ios/iosimagepicker.cpp]: encapsulate -> move to private and rename with starting "m". (mOwner,..) create setter,getter

File: app/ios/iosimagepicker.cpp [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+    QImage image;
+    QString absoluteImagePath;
+
+    /// <#Description#>
```
Comment (PeterPetrik) [inline app/ios/iosimagepicker.cpp]: use doxygen comments

File: app/ios/iosimagepicker.cpp [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+    void run()
+    {
+      if ( absoluteImagePath.isNull() )
+      {
+        absoluteImagePath = QString( "%1/%2.jpg" ).arg( owner->targetPath(), QDateTime::currentDateTime().toString( QStringLiteral( "yyMMdd-hhmmss" ) ) );
```
Comment (PeterPetrik) [inline app/ios/iosimagepicker.cpp]: line too long , split

File: app/ios/iosimagepicker.cpp [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+        QMetaObject::invokeMethod( owner.data(), "endSave", Qt::QueuedConnection,
+                                   Q_ARG( QString, absoluteImagePath ) );
+      }
+    }
+};
+
+IOSImagePicker::IOSImagePicker( QObject *parent ) : QObject( parent )
+{
+  mSourceType = PhotoLibrary;
+  mStatus = Null;
```
Comment (PeterPetrik) [inline app/ios/iosimagepicker.cpp]: Null?

File: app/ios/iosimagepicker.cpp [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+  mBusy = false;
+}
+
+IOSImagePicker::~IOSImagePicker()
```
Comment (PeterPetrik) [inline app/ios/iosimagepicker.cpp]: = default

File: app/ios/iosimagepicker.cpp [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+  saver->owner = this;
+  saver->absoluteImagePath = fileName;
+  saver->image = mImage;
+
+  QThreadPool::globalInstance()->start( saver );
+}
+
+void IOSImagePicker::saveImage()
+{
+  save( QString() );
```
Comment (PeterPetrik) [inline app/ios/iosimagepicker.cpp]: is this right?

File: app/ios/iosimagepicker.cpp [hunk 19]
```diff
@@ -0,0 +1,265 @@
...
+
+QImage IOSImagePicker::image() const
+{
+  return mImage;
+}
+
+void IOSImagePicker::setImage( const QImage &image )
+{
+  mImage = image;
+  if ( true )
```
Comment (PeterPetrik) [inline app/ios/iosimagepicker.cpp]: if true...?

File: app/ios/iosimagepicker.cpp [hunk 20]
```diff
@@ -0,0 +1,265 @@
...
+  QISystemDispatcher *system = QISystemDispatcher::instance();
+
+  QVariantMap data;
+  data["active"] = mBusy;
+
+  system->dispatch( "imagePickerControllerSetIndicator", data );
+  emit busyChanged();
+}
+
+
```
Comment (PeterPetrik) [inline app/ios/iosimagepicker.cpp]: too many new lines

File: app/ios/iosimagepicker.h [hunk 21]
```diff
@@ -0,0 +1,111 @@
+#ifndef QIIMAGEPICKER_H
```
Comment (PeterPetrik) [inline app/ios/iosimagepicker.h]: - add copyright header
- add to ios.pri sources?

File: app/ios/iosimagepicker.h [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+
+#include <QObject>
+#include <QImage>
+
+#include <QVariant>
+#include <QString>
+#include <QMap>
+#include <QVariantMap>
+
+/// IOSImagePicker provides a simple interface to access camera and camera roll via the UIImagePickerController
```
Comment (PeterPetrik) [inline app/ios/iosimagepicker.h]: use doxy comments

Comment (PeterPetrik) [reply app/ios/iosimagepicker.h]: add some docs to functions where needed

File: app/ios/iossystemdispatcher.cpp [hunk 23]
```diff
@@ -0,0 +1,50 @@
+#include <QCoreApplication>
```
Comment (PeterPetrik) [inline app/ios/iossystemdispatcher.cpp]: - add copyright header
- add to ios.pri sources?

File: app/ios/iossystemdispatcher.cpp [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  return res;
+}
+
+bool QISystemDispatcher::addListener( QString name, bool ( *func )( QVariantMap & ) )
+{
+  if ( handlers.contains( name ) )
+  {
+    qWarning() << QString( "%s is already registered" ).arg( name );
```
Comment (PeterPetrik) [inline app/ios/iossystemdispatcher.cpp]: do we use qWarnings()?

File: app/ios/iossystemdispatcher.h [hunk 25]
```diff
@@ -0,0 +1,42 @@
+#pragma once
+#include <QObject>
+#include <QVariantMap>
+
+/// QISystemMessenger provides an simple async messaging interface between C/C++/QML and Objective-C source code.
```
Comment (PeterPetrik) [inline app/ios/iossystemdispatcher.h]: use doxy comments

File: app/ios/iossystemdispatcher.h [hunk 26]
```diff
@@ -0,0 +1,42 @@
+#pragma once
```
Comment (PeterPetrik) [inline app/ios/iossystemdispatcher.h]: - add copyright header
- add to ios.pri sources?

why pragma?

File: app/ios/iossystemdispatcher.h [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+
+class QISystemDispatcher : public QObject
+{
+    Q_OBJECT
+
+  public:
+    static QISystemDispatcher *instance();
+
+    /// Deliver a message
+    /** If there has a registered listener function , it will return TRUE. Otherwise, it will return FALSE.
```
Comment (PeterPetrik) [inline app/ios/iossystemdispatcher.h]: fix docs in all functions below

File: app/ios/iossystemutils.mm [hunk 28]
```diff
@@ -0,0 +1,396 @@
+#include <QCoreApplication>
```
Comment (PeterPetrik) [inline app/ios/iossystemutils.mm]: - add copyright header

File: app/ios/iossystemutils.mm [hunk 29]
```diff
@@ -0,0 +1,396 @@
...
+
+bool imagePickerControllerDismiss(QVariantMap& data) {
+    Q_UNUSED(data);
+    if (!imagePickerController)
+        return false;
+
+    bool animated = data["animated"].toBool();
+
+    [imagePickerController dismissViewControllerAnimated:animated completion:NULL];
+//    [imagePickerController release];
```
Comment (PeterPetrik) [inline app/ios/iossystemutils.mm]: remove commented out code?

File: app/ios/iossystemutils.mm [hunk 30]
```diff
@@ -0,0 +1,396 @@
...
+
+static bool activityIndicatorStopAnimation(QVariantMap& data) {
+    Q_UNUSED(data);
+
+    if (!activityIndicator) {
+        return false;
+    }
+
+    [activityIndicator stopAnimating];
+//    [activityIndicator release];
```
Comment (PeterPetrik) [inline app/ios/iossystemutils.mm]: commented out code

File: app/ios/iosviewdelegate.h [hunk 31]
```diff
@@ -0,0 +1,24 @@
+//
```
Comment (PeterPetrik) [inline app/ios/iosviewdelegate.h]: - add copyright header
- add to ios.pri sources?

File: app/ios/iosviewdelegate.mm [hunk 32]
```diff
@@ -0,0 +1,45 @@
+#include <QtCore>
```
Comment (PeterPetrik) [inline app/ios/iosviewdelegate.mm]: - add copyright header

**Comment by sklencar** [comment]
Obsolate due to https://github.com/lutraconsulting/input/pull/503

**Pull request closed without merging**